litex/soc/software/bios/commands/cmd_dram.c

```diff
@@ -0,0 +1,309 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <generated/csr.h>
+
+#include "../command.h"
+#include "../helpers.h"
+#include "../sdram.h"
+
+#ifdef CSR_SDRAM_BASE
+/**************************************
+ *            Memory read             *
+ **************************************
+ */
+static void sdrrow_handler(int nb_params, char **params)
+{
+	char *c;
+	unsigned int row;
+
+	printf("\n");
+	if (nb_params < 1) {
+		sdrrow(0);
+		printf("Precharged");
+	}
+	
+	row = strtoul(params[0], &c, 0);
+	if (*c != 0) {
+		printf("Incorrect row");
+		return;
+	}
+
+	sdrrow(row);
+	printf("Activated row %d", row);
+}
+
+struct command_struct cmd_sdrrow =
+{
+	.func = sdrrow_handler,
+	.name = "sdrrow",
+	.help = "Precharge/Activate row",
+};
+
+define_command(cmd_sdrrow);
+
+/**************************************
+ *          SDRAM control             *
+ **************************************
+ */
+static void sdrsw_handler(int nb_params, char **params)
+{
+	printf("\n");
+	sdrsw();
+}
+
+struct command_struct cmd_sdrsw =
+{
+	.func = sdrsw_handler,
+	.name = "sdrsw",
+	.help = "Gives SDRAM control to SW",
+};
+
+define_command(cmd_sdrsw);
+
+static void sdrhw_handler(int nb_params, char **params)
+{
+	printf("\n");
+	sdrhw();
+}
+
+struct command_struct cmd_sdrhw =
+{
+	.func = sdrhw_handler,
+	.name = "sdrhw",
+	.help = "Gives SDRAM control to HW",
+};
+
+define_command(cmd_sdrhw);
```

Comment:
Reading this now makes me think I could save space if .func would point directly to sdrhw()